### PR TITLE
Bus-mapping for opcode `calldatacopy`

### PR DESCRIPTION
**Pull request by silathdiir**
Close https://github.com/appliedzkp/zkevm-circuits/issues/309

### Summary

1. Add `ExecState` to bus-mapping `ExecStep` to support Virtual step BeginTx, EndTx and CopyToMemory.
2. Add function `new_step` to `CircuitInputStateRef` for multi-step support.
3. Implement `CallDataCopy` bus-mapping only for root call. Will send another PR for internal call of CallDataCopy bus-mapping when PR https://github.com/appliedzkp/zkevm-circuits/pull/278 is merged.

### CODE REVIEW

**Comment by icemelon** [comment]
I suggest that we should merge this PR even though `calldatacopy` with internal call is not supported yet, because it changes the `gen_assoicated_op` function signature and could affect other bus-mapping PRs.

File: bus-mapping/src/circuit_input_builder.rs
Comment (ed255) [inline bus-mapping/src/circuit_input_builder.rs:734]: Related to my previous comment: maybe here we also need a new function to create a virtual step.

**Review by ed255 — APPROVED** [review]
Great work!  Thanks for working on this :)

I've added a few comments with minor suggestions, but overall everything looks good!

File: bus-mapping/src/evm/opcodes/calldatacopy.rs
Comment (icemelon) [reply bus-mapping/src/evm/opcodes/calldatacopy.rs]: fixed

File: bus-mapping/src/circuit_input_builder.rs
Comment (icemelon) [reply bus-mapping/src/circuit_input_builder.rs]: found out that we don't need to provide `selectors` in the witness, and in fact, it could be derived in the witness assignment. So removed from the struct

File: bus-mapping/src/evm/opcodes.rs
Comment (icemelon) [reply bus-mapping/src/evm/opcodes.rs:55]: updated

File: bus-mapping/src/evm/opcodes/number.rs
Comment (icemelon) [reply bus-mapping/src/evm/opcodes/number.rs]: probably due to rebase. revert the change

File: bus-mapping/src/evm/opcodes.rs
Comment (icemelon) [reply bus-mapping/src/evm/opcodes.rs]: I added extra functions `new_begin_tx_step` and `new_end_tx_step` in the `CircuitInputStateRef`. Because the logic of creating begin_tx step and end_tx step are rather different, it makes more sense, imo, that just provides a special API for creating these two special steps.

File: bus-mapping/src/evm/opcodes/calldatacopy.rs
Comment (icemelon) [reply bus-mapping/src/evm/opcodes/calldatacopy.rs:140]: I think this part should be okay with `state.new_step` since we only need to change the `exec_state` to `ExecState::CopyToMemory`.

Comment (ed255) [reply bus-mapping/src/evm/opcodes/calldatacopy.rs:140]: Ok, seems reasonable!

**Comment by ed255** [comment]
Let's wait for a second review and afterwards we can merge this!

**Pull request merged**
